File: python/include/partition.py
from typing import List
import heapq
import copy
import math

def LS(arr: List[int], n: int) -> List[List[int]]:
   """
   Partition a list of integers into n groups using a list scheduling algorithm.

   The algorithm assigns each element to the subset with the current smallest sum.
   This produces a reasonably balanced distribution, but not necessarily optimal.

   Parameters
   ----------
   arr : List[int]
      The list of integers to partition.
   n : int
      The number of groups to create.

   Returns
   -------
   List[List[int]]
      A list containing n groups (each subset is a list of integers).
   """
   if n <= 0:
      raise ValueError("n must be a positive integer")
   if n == 1:
      return [arr]

   groups = [[] for _ in range(n)]
   pq = [(0, i) for i in range(n)]
   heapq.heapify(pq)

   for x in arr:
      sum, i = heapq.heappop(pq)
      groups[i].append(x)
      heapq.heappush(pq, (sum + x, i))

   return groups

def LPT(arr: List[int], n: int) -> List[List[int]]:
   """
   Partition a list of integers into k groups using the LPT (Longest Processing Time) heuristic.

   The algorithm sorts the list in descending order and then applies the greedy algorithm.
   This often produces a better balance than greedy alone.

   Parameters
   ----------
   arr : List[int]
      The list of integers to partition.
   n : int
      The number of groups to create.

   Returns
   -------
   List[List[int]]
      A list containing k groups (each subset is a list of integers).
   """
   if n <= 0:
      raise ValueError("K must be a positive integer")
   if n == 1:
      return [arr]

   return LS(sorted(arr, reverse=True), n)

def MULTIFIT(arr: List[int], n: int, k: int = 7) -> List[List[int]]:
   """
   Partition a list of integers into k groups using the MULTIFIT (MULTI-FIT) heuristic.

   The algorithm first sorts the list in descending order and then iteratively constructs the final k groups.
   In each iteration, it computes the average capacity of the groups and delegates to the internal FFD
   procedure to construct new groups. If the number of groups exceeds k, it updates the lower bound for the next iteration.
   Otherwise, it updates the upper bound and stores the current groups as the best solution.

   Parameters
   ----------
   arr : List[int]
      The list of integers to partition.
   n : int
      The number of groups to create.
   k : int
      The number of iterations to run the algorithm (default is 7).

   Returns
   -------
   List[List[int]]
      A list containing k groups (each subset is a list of integers).
   """
   if n <= 0:
      raise ValueError("K must be a positive integer")
   if n == 1:
      return [arr]
   
   arr = sorted(arr, reverse=True)
   
   sum = 0
   max_val = 0
   for x in arr:
      sum += x
      max_val = max(max_val, x)

   lowerBound = max(max_val, sum // n)
   upperBound = max(max_val, 2 * sum // n)

   bestGroups = []
   
   for _ in range(k):
      capacity = (lowerBound + upperBound) // 2
      groups = FFD(arr, capacity)

      if len(groups) > n:
         lowerBound = capacity
      else:
         bestGroups = groups
         upperBound = capacity
   
   return bestGroups

def FFD(arr: List[int], capacity: int) -> List[List[int]]:
   """
   First-Fit Decreasing (FFD) algorithm to partition a list of integers into groups
   with a given capacity.

   The algorithm assigns each element to the subset with the current smallest remaining
   capacity, in order to balance the groups.

   Parameters
   ----------
   arr : List[int]
      The list of integers to partition.
   capacity : int
      The capacity of each group.

   Returns
   -------
   List[List[int]]
      A list containing the partitioned groups of the array.
   """
   groups = []
   remaining = []
   
   for x in arr:
      placed = False
      for i in range(len(groups)):
         if remaining[i] >= x:
            groups[i].append(x)
            remaining[i] -= x
            placed = True
            break
      if not placed:
         groups.append([x])
         remaining.append(capacity - x)
   
   return groups

def CGA(arr: List[int], n: int) -> List[List[int]]:
   """
   Complete Greedy Algorithm (CGA) to partition a list of integers into n groups.
   
   The algorithm first applies the Longest Processing Time (LPT) heuristic to find an initial solution.
   Then, it uses backtracking to explore all possible solutions and find the best one.
   
   Parameters
   ----------
   arr : List[int]
      The list of integers to partition.
   n : int
      The number of groups to create.
   
   Returns
   -------
   List[List[int]]
      A list containing the best partitioned groups of the array.
   """
   # Check if n is valid
   if n <= 0:
      raise ValueError("K must be a positive integer")
   if n == 1:
      return [arr]
   
   # Get one solution (heurística inicial)
   groups = LPT(arr, n)
   
   # get sums of groups
   sums = list(map(sum, groups))

   # Get makespan
   makespan = max(sums)

   # Get makespan lowerbound
   total = sum(sums)
   lowerbound = math.ceil(float(total) / n)
   
   # Get best solution
   if lowerbound < makespan:
      groupsSums = [0] * n
      actualGroups = [[] for _ in range(n)]
      makespan, groups = CGABacktracking(
         arr, 
         actualGroups, 
         groupsSums, 
         makespan, 
         lowerbound,
         copy.deepcopy(groups),
         0
      )
   
   # Return best solution
   return groups


def CGABacktracking(
   arr: List[int], 
   groups: List[List[int]], 
   groupSums: List[int], 
   makespan: int, 
   lowerbound: int,
   groupsCandidate: List[List[int]], 
   i: int
):
   """
   Backtracking algorithm to find the optimal partition of the array
   into n groups using a Complete Greedy Algorithm (CGA) approach.

   The algorithm iteratively assigns each element to the group with the
   current minimum sum, in order to balance the groups. It uses a
   backtracking approach to explore all possible combinations of groups.

   Parameters
   ----------
   arr : List[int]
      The list of integers to partition.
   groups : List[List[int]]
      The current partition of the array.
   groupSums : List[int]
      The sum of each group in the current partition.
   makespan : int
      The maximum difference between the sums of the groups.
   lowerbound : int
      The lower bound of the makespan.
   groupsCandidate : List[List[int]]
      The current candidate partition.
   i : int
      The current index in the array.

   Returns
   -------
   int
      The maximum difference between the sums of the groups.
   List[List[int]]
      The current candidate partition.
   """
   # Base case
   if i == len(arr):
      currentMax = max(groupSums)
      # Update
      if currentMax < makespan:
         makespan = currentMax
         groupsCandidate = copy.deepcopy(groups)
      return makespan, groupsCandidate

   # Sort groups by carga atual
   groupsIndices = list(range(len(groups)))
   groupsIndices.sort(key=lambda idx: groupSums[idx])
   
   # Tried sums
   triedSums = set()
   
   # Backtracking
   for j in groupsIndices:
      # Not already tried
      if groupSums[j] in triedSums:
         continue
      triedSums.add(groupSums[j])
      
      # Evaluation
      groupSums[j] += arr[i]
      currentMax = max(groupSums)

      # Uperbound prune
      if currentMax < makespan:
         # Recursion
         groups[j].append(arr[i])
         makespan, groupsCandidate = CGABacktracking(
            arr, groups, groupSums, makespan, lowerbound, groupsCandidate, i + 1
         )
         groups[j].pop() # Backtrack
         
         # Lowerbound prune
         if makespan == lowerbound:
            return makespan, groupsCandidate
      
      # Backtrack 
      groupSums[j] -= arr[i]
   
   return makespan, groupsCandidate

File: java/src/partition/Partition.java
package partition;

import java.util.*;

/**
 * Utility class for partitioning a list of integers into {@code k} groups
 * using a greedy (approximate) strategy.
 *
 * <p>The algorithm works as follows:
 * each element from the input list is assigned to the group
 * that currently has the smallest sum, in order to balance the groups.</p>
 *
 * <p>Example usage:</p>
 * <pre>{@code
 * List<Integer> numbers = List.of(10, 20, 30, 40, 50);
 * List<List<Integer>> groups = Partition.greedy(numbers, 3);
 * }</pre>
 *
 * The result will be a list of 3 groups with approximately balanced sums.
 */
public class Partition {

    /**
     * Partitions a list of integers into {@code n} groups using a List Scheduling approach.
     *
     * <p>The method iteratively assigns each integer to the group
     * with the current minimum sum. This does not guarantee an optimal solution,
     * but it produces a reasonably balanced distribution in practice.</p>
     *
     * @param arr the list of integers to be partitioned
     * @param n the number of groups (must be greater than zero)
     * @return a list containing {@code n} sublists, each representing a group
     * @throws IllegalArgumentException if {@code n <= 0}
     */
    public static List<List<Integer>> LS(List<Integer> arr, int n) throws IllegalArgumentException {
        if (n <= 0) {
            throw new IllegalArgumentException("K must be a positive integer");
        } else if (n == 1) {
            return List.of(new ArrayList<>(arr));
        }

        // Initialize groups and their sums
        List<List<Integer>> groups = new ArrayList<>(n);
        var pq = new PriorityQueue<int[]>(Comparator.comparingInt(a -> a[0]));

        for (int i = 0; i < n; i++) {
            groups.add(new ArrayList<>());
            pq.add(new int[]{0, i});
        }

        // Distribute elements greedily
        for (int x : arr) {
            int[] top = pq.poll();
            assert top != null;
            int sum = top[0], idx = top[1];
            groups.get(idx).add(x);
            pq.add(new int[]{sum + x, idx});
        }

        return groups;
    }

    /**
     * Applies the Longest Processing Time (LPT) heuristic to partition a list of integers
     * into {@code n} groups.
     *
     * <p>The method first sorts the list in descending order (largest numbers first),
     * then delegates to the {@link #LS(List, int)} method to distribute the elements.
     * This heuristic is commonly used in load balancing and scheduling problems.</p>
     *
     * @param arr the list of integers to be partitioned
     * @param n the number of groups (must be greater than zero)
     * @return a list containing {@code n} sublists, each representing a group
     * @throws IllegalArgumentException if {@code n <= 0}
     */
    public static List<List<Integer>> LPT(List<Integer> arr, int n) throws IllegalArgumentException {
        if (n <= 0) {
            throw new IllegalArgumentException("n must be a positive integer");
        } else if (n == 1) {
            return List.of(new ArrayList<>(arr));
        }

        // Sort in descending order (largest elements first)
        arr.sort(Comparator.reverseOrder());

        // Delegate to the greedy algorithm
        return LS(arr, n);
    }

    /**
     * Applies the MULTIFIT heuristic to partition a list of integers into {@code n} groups.
     *
     * <p>This method provides a simplified interface that uses a default of seven binary
     * search iterations to estimate the optimal bin capacity. It delegates to the
     * {@link #MULTIFIT(List, int, int)} method for the actual computation.</p>
     *
     * <p>The MULTIFIT algorithm is a hybrid approach that combines binary search with
     * the First-Fit Decreasing (FFD) heuristic, often used in bin packing and load
     * balancing problems. It aims to find the smallest possible bin capacity such that
     * all elements can be packed into {@code n} bins.</p>
     *
     * @param arr the list of integers to be partitioned
     * @param n the number of groups (bins) to create (must be greater than zero)
     * @return a list of {@code n} sublists, each representing a group
     * @throws IllegalArgumentException if {@code n <= 0}
     * @see #MULTIFIT(List, int, int)
     * @see #FFD(List, int)
     */
    public static List<List<Integer>> MULTIFIT(List<Integer> arr, int n) throws IllegalArgumentException {
        return MULTIFIT(arr, n, 7);
    }

    /**
     * Applies the MULTIFIT heuristic with a configurable number of binary search iterations.
     *
     * <p>The MULTIFIT algorithm iteratively adjusts the bin capacity between lower and
     * upper bounds. In each iteration, it applies the {@link #FFD(List, int)} method
     * (First-Fit Decreasing) to test whether all items can fit into {@code n} bins
     * with the current capacity. The search continues until the best feasible
     * capacity is found.</p>
     *
     * <p>The algorithm was first introduced by Coffman, Garey, and Johnson (1978)
     * and is known for providing a good approximation for the bin packing problem.</p>
     *
     * @param arr the list of integers to be partitioned
     * @param n the number of bins to use (must be greater than zero)
     * @param k the number of binary search iterations (default is typically 7)
     * @return a list of sublists, where each sublist represents one bin or group
     * @throws IllegalArgumentException if {@code n <= 0}
     * @see #FFD(List, int)
     */
    public static List<List<Integer>> MULTIFIT(List<Integer> arr, int n, int k) throws IllegalArgumentException {
        if (n <= 0) {
            throw new IllegalArgumentException("n must be a positive integer");
        } else if (n == 1) {
            return List.of(new ArrayList<>(arr));
        }

        arr.sort(Comparator.reverseOrder());

        int sum = 0, max = 0;
        for (int x : arr) {
            sum += x;
            max = Math.max(max, x);
        }

        int lowerBound = Math.max(max, sum / n);
        int upperBound = Math.max(max, 2 * sum / n);

        List<List<Integer>> bestGroups = new ArrayList<>();

        for (int i = 0; i < k; i++) {
            int capacity = (lowerBound + upperBound) / 2;
            var groups = FFD(arr, capacity);

            if (groups.size() > n) {
                lowerBound = capacity;
            } else {
                bestGroups = groups;
                upperBound = capacity;
            }
        }

        return bestGroups;
    }

    /**
     * Applies the First-Fit Decreasing (FFD) heuristic to partition a list of integers
     * into groups with a given capacity.
     *
     * <p>The algorithm first considers each item in descending order (as provided),
     * and assigns it to the first bin that has enough remaining capacity.
     * If no such bin exists, a new bin is created. This greedy approach
     * is efficient and widely used for bin packing problems.</p>
     *
     * @param arr the list of integers to be partitioned
     * @param capacity the maximum capacity allowed per bin
     * @return a list of sublists, each representing a bin containing one or more items
     * @see #MULTIFIT(List, int, int)
     */
    public static List<List<Integer>> FFD(List<Integer> arr, int capacity) {
        record Bin(int remaining, int idx) implements Comparable<Bin> {
            @Override
            public int compareTo(Bin o) {
                return Integer.compare(remaining, o.remaining);
            }
        }

        List<List<Integer>> groups = new ArrayList<>();
        TreeSet<Bin> bins = new TreeSet<>();

        for (int x : arr) {
            Bin it = bins.ceiling(new Bin(x, 0));

            if (it == null) {
                groups.add(new ArrayList<>(x));
                bins.add(new Bin(capacity - x, groups.size() - 1));
            } else {
                groups.get(it.idx).add(x);
                bins.remove(it);
                bins.add(new Bin(it.remaining - x, it.idx));
            }
        }

        return groups;
    }
}
